src/Apply.hs

```diff
@@ -113,5 +114,23 @@ classify xs i = let s = foldl' (f i) (ideaSeverity i) xs in s `seq` i{ideaSeveri
         f :: Idea -> Severity -> Classify -> Severity
         f i r c | classifyHint c ~~= ideaHint i && classifyModule c ~= ideaModule i && classifyDecl c ~= ideaDecl i = classifySeverity c
                 | otherwise = r
-        x ~= y = x == "" || x `elem` y
+        x ~= y = x == "" || x `elem` y || any (wildcardMatch x) y
         x  ~~= y = x == "" || x == y || ((x ++ ":") `isPrefixOf` y)
+
+-- | Returns true if the pattern matches the string. For example:
+--
+-- >>> let isSpec = wildcardMatch "**.*Spec"
+-- >>> isSpec "Example"
+-- False
+-- >>> isSpec "ExampleSpec"
+-- True
+-- >>> isSpec "Namespaced.ExampleSpec"
+-- True
+-- >>> isSpec "Deeply.Nested.ExampleSpec"
+-- True
+--
+-- See this issue for details: <https://github.com/ndmitchell/hlint/issues/402>.
+wildcardMatch :: FilePattern -> String -> Bool
+wildcardMatch p m =
+    let f x = if x == '.' then '/' else x
+    in '*' `elem` p && fmap f p ?== fmap f m
```

Comment:
Is the `*` test is a "cheap test" to check if it works? If yes, but has no functional difference, a comment would be good. If no, and it has a real semantic difference, what is it?

Reply:
Yes, it was meant as a cheap test to see if we should even bother trying a wildcard match. However now that I think about it we should be able to combine the regular and wildcard matchers. A regular matcher is just a wildcard that doesn't have any `*`s in it.